File: framework/lib/chrome.dart
part of uxml;

/**
 * Creates element tree for a control's visual look and behaviour.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class Chrome {
  String _id;
  Object _type;
  CreateElementsDelegate _createElements;
  List<Effect> _effects;

  /**
  * Constructor.
  */
  Chrome(String id, Object type, CreateElementsDelegate foo) {
    _id = id;
    _type = type;
    _createElements = foo;
    _effects = null;
  }

  /**
   * Returns id.
   */
  String get id => _id;

  /**
   * Returns effects collection. Lazily creates one if empty.
   */
  List<Effect> get effects {
    if (_effects == null) {
      _effects = <Effect>[];
    }
    return _effects;
  }

  /**
   * Returns target type of chrome.
   */
  Object get type => _type;

  /**
   * Overridable function that creates the element tree and applies effects.
   */
  UIElement applyToTarget(UIElement targetElement) {
    UIElement newChromeTree = null;
    if (_createElements != null) {
      newChromeTree = _createElements(targetElement);
    }
    if (effects != null && effects.length != 0) {
      for (int i = 0; i < effects.length; i++) {
        Effect effect = effects[i];
        Effect newEffect = effect.clone();
        if (newEffect.targetElement == null) {
          newEffect.targetElement = targetElement;
        }
        targetElement.effects.add(newEffect);
      }
    }
    return newChromeTree;
  }

  /**
   * Sets the default value of a property defined on it's chrome if it
   * has not been changed yet.
   */
  static void applyProperty(UxmlElement targetObject,
                            PropertyDefinition key, Object value) {
    if (!targetObject.overridesProperty(key)) {
      targetObject.setProperty(key, value);
    }
  }
}

typedef UIElement CreateElementsDelegate(UIElement target);

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/Chrome.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.effects.Effect;
import com.hello.uxml.tools.framework.events.ChromeHandler;

import java.util.ArrayList;
import java.util.List;

/**
 * Creates element tree for a control's visual look and behaviour.
 *
 * @author ferhat@(Ferhat Buyukkokten)
 */
public class Chrome extends UxmlElement {

  private Class<?> targetType;
  private List<Effect> effects;
  private ChromeHandler createElementHandler;

  /**
   * Constructor.
   */
  public Chrome() {
  }

  /**
   * Constructor.
   */
  public Chrome(String id, Class<?> targetType, ChromeHandler createElementHandler) {
    this.setId(id);
    this.targetType = targetType;
    this.createElementHandler = createElementHandler;
  }

  /**
   * Sets/returns target type of chrome.
   */
  public Class<?> getTargetType() {
    return targetType;
  }

  public void setTargetType(Class<?> targetType) {
    this.targetType = targetType;
  }

  public List<Effect> getEffects() {
    if (effects == null) {
      effects = new ArrayList<Effect>();
    }
    return effects;
  }

  /**
   * Overridable function that creates the element tree.
   */
  public UIElement apply(UIElement targetElement) {
    if (createElementHandler != null) {
      return createElementHandler.createElements(targetElement);
    }
    ContentContainer contentContainer = new ContentContainer();
    return contentContainer;
  }
}
